### PR TITLE
Always allow full filesystem access to LSP.

### PR DESCRIPTION
**Pull request by chriseth**
Maybe we need a different solution for Windows? Something like a flag to just disable the check inside the file reader?

### CODE REVIEW

**Comment by ekpyron** [comment]
@christianparpart For CI to pass this will need a rebase to current ``develop``.

File: libsolidity/lsp/FileRepository.h [hunk 1]
```diff
@@ -29,7 +29,10 @@ class FileRepository
 {
 public:
 	explicit FileRepository(boost::filesystem::path const& _basePath):
-		m_fileReader(_basePath) {}
+		m_fileReader(_basePath)
+	{
+		m_fileReader.allowDirectory(boost::filesystem::path("/"));
```
Comment (chriseth) [inline libsolidity/lsp/FileRepository.h]: Does this work on windows?

**Review by Aivolt1 — DISMISSED** [review]

File: libsolidity/lsp/FileRepository.h (hunk 1 above)
Comment (christianparpart) [reply libsolidity/lsp/FileRepository.h]: Yes, with my changes added to this PR: All tests pass (locally).

File: test/lsp.py [hunk 2]
```diff
@@ -180,6 +177,10 @@ def __init__(self):
         args = create_cli_parser().parse_args()
         self.solc_path = args.solc_path
         self.project_root_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
+        if os.name == 'nt':
```
Comment (chriseth) [inline test/lsp.py]: What about using `as_uri()` from `pathlib`?

File: test/lsp.py [hunk 3]
```diff
@@ -60,10 +60,7 @@ def receive_message(self) -> Union[None, dict]:
                 # server quit
                 return None
             line = line.decode("utf-8")
-            if not line.endswith("\r\n"):
```
Comment (chriseth) [inline test/lsp.py:146]: Why did you remove this check?

File: libsolidity/lsp/Transport.cpp [hunk 4]
```diff
@@ -107,8 +107,15 @@ void IOStreamTransport::send(Json::Value _json, MessageID _id)
 
 	string const jsonString = solidity::util::jsonCompactPrint(_json);
 
-	m_output << "Content-Length: " << jsonString.size() << "\r\n";
-	m_output << "\r\n";
+#if defined(_WIN32)
```
Comment (chriseth) [inline libsolidity/lsp/Transport.cpp]: I don't think this is a proper solution. I saw some posts mentioning freopen and others `setmode`: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setmode?view=msvc-170&viewFallbackFrom=vs-2019

File: libsolidity/lsp/FileRepository.cpp [hunk 5]
```diff
@@ -42,7 +42,16 @@ string FileRepository::sourceUnitNameToClientPath(string const& _sourceUnitName)
 	else if (_sourceUnitName.find("file://") == 0)
 		return _sourceUnitName;
 	else
-		return "file://" + (m_fileReader.basePath() / _sourceUnitName).generic_string();
+	{
+		auto basePath = m_fileReader.basePath();
+#if defined(_WIN32)
+		// Ensure basePath does contain Windows drive letter to fully qualify the path.
+		if (!basePath.has_root_name())
+			basePath = (boost::filesystem::current_path().root_path() / basePath).normalize();
```
Comment (chriseth) [inline libsolidity/lsp/FileRepository.cpp]: I don't think we should use the current path.

File: test/lsp.py (hunk 2 above)
Comment (christianparpart) [reply test/lsp.py]: I did not know about that. will be trying.

File: test/lsp.py (hunk 3 above)
Comment (christianparpart) [reply test/lsp.py:146]: the rstrip does also remove CRLF but more relaxed, such that it would have allowed a simple LF too. I've reverted it though.

File: libsolidity/lsp/Transport.cpp [hunk 6]
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
...
-		if (delimiterPos == string::npos)
-			return nullopt;
+std::string StdioTransport::readBytes(size_t _byteCount)
+{
+	std::string buffer;
+	buffer.resize(_byteCount);
+	auto const n = fread(buffer.data(), 1, _byteCount, stdin);
+	if (n < _byteCount)
+		buffer.resize(n);
+	lspDebug(fmt::format("StdioTransport.readBytes({};{};{}): \"{}\"",
```
Comment (chriseth) [inline libsolidity/lsp/Transport.cpp]: coding style

File: libsolidity/lsp/Transport.cpp [hunk 7]
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
...
-			boost::trim_copy(value)
-		).second)
-			return nullopt;
-	}
-	return {move(headers)};
+std::string StdioTransport::getline()
+{
+	char* lineBuffer = nullptr;
+	size_t numBytesRead = 0;
+	auto const getlineResult = ::getline(&lineBuffer, &numBytesRead, stdin);
```
Comment (chriseth) [inline libsolidity/lsp/Transport.cpp]: Can you use an explicit type please?

File: libsolidity/lsp/Transport.cpp [hunk 8]
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
...
+	auto const ensureFreedBuffer = ScopeGuard{ [&]() { free(lineBuffer); }};
```
Comment (chriseth) [inline libsolidity/lsp/Transport.cpp]:
```suggestion
	ScopeGuard const ensureFreedBuffer{[&]() { free(lineBuffer); }};
```

File: libsolidity/lsp/Transport.cpp [hunk 9]
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
...
+	if (getlineResult > 0)
+		return std::string(lineBuffer, static_cast<size_t>(getlineResult));
+	else
+		return {};
+}
+
+void StdioTransport::writeBytes(std::string_view _data)
+{
+	auto const bytesWritten = fwrite(_data.data(), 1, _data.size(), stdout);
```
Comment (chriseth) [inline libsolidity/lsp/Transport.cpp:213]: Please use an explicit type

**Comment by leonardoalt** [comment]
Is this PR waiting for fixes or reviews? @chriseth @christianparpart

**Comment by christianparpart** [comment]
> Is this PR waiting for fixes or reviews? @chriseth @christianparpart

yes, please. :)

File: libsolidity/lsp/FileRepository.cpp [hunk 10]
```diff
@@ -17,48 +17,129 @@
...
+
+	try
+	{
+		// File was read already. Use local store.
+		if (m_sourceCodes.count(_sourceUnitName))
+			return ReadCallback::Result{true, m_sourceCodes.at(_sourceUnitName)};
+
+		string const strippedSourceUnitName = stripFileUriSchemePrefix(_sourceUnitName);
+
+		if (boost::filesystem::path(strippedSourceUnitName).has_root_path() &&
```
Comment (Marenz) [inline libsolidity/lsp/FileRepository.cpp]: `if (` should be on its own

**Review by Marenz — CHANGES_REQUESTED** [review]
Found a tiny styling problem, but pylint also demands fixes.

File: libsolidity/lsp/Transport.cpp [hunk 11]
```diff
@@ -75,79 +69,156 @@ optional<Json::Value> IOStreamTransport::receive()
...
+		if (boost::trim_copy(line).empty())
+			break;
+
+		auto const delimiterPos = line.find(':');
+		if (delimiterPos == string::npos)
+			return nullopt;
+
+		auto const name = boost::to_lower_copy(line.substr(0, delimiterPos));
+		auto const value = line.substr(delimiterPos + 1);
+		if (!headers.emplace(
```
Comment (Marenz) [inline libsolidity/lsp/Transport.cpp]: `if (` should be on its own

File: test/lsp.py [hunk 12]
```diff
@@ -1,26 +1,57 @@
...
+    """
+    Gets a single character from stdin without line-buffering.
+    """
+    if os.name == 'nt':
+        # pragma pylint: disable=import-error
+        return msvcrt.getch().decode("utf-8")
+    else:
+        # Turn off user input buffering so we get the input immediately,
+        # not only after a line break
+        tty.setcbreak(sys.stdin.fileno())
```
Comment (Marenz) [inline test/lsp.py]: This doesn't need to be set every time we want a char. Once in the life-time of the program is enough.

**Review by Marenz — CHANGES_REQUESTED** [review]

File: libsolidity/lsp/Transport.cpp (hunk 11 above)
Comment (christianparpart) [reply libsolidity/lsp/Transport.cpp]: This is just moved code, not new one. I change change however. :)

File: test/lsp.py (hunk 12 above)
Comment (christianparpart) [reply test/lsp.py]: Exactly, but that way I kept the logic more closely together without over-complifying something simple.

**Review by Marenz — DISMISSED** [review]

**Review by Marenz — APPROVED** [review]

File: libsolidity/lsp/Transport.h [hunk 13]
```diff
@@ -119,27 +143,34 @@ class IOStreamTransport: public Transport
...
+	void writeBytes(std::string_view _data) override;
+	void flushOutput() override;
 
 private:
 	std::istream& m_input;
 	std::ostream& m_output;
 };
 
+/**
+ * Standard I/O transport Layer utilizing stdin/stdout for communication.
```
Comment (chriseth) [inline libsolidity/lsp/Transport.h:160]: The changes in these files do not look like they are about allowing full filesystem access. Please put them in their own PR in the future.

Comment (christianparpart) [reply libsolidity/lsp/Transport.h:160]: I was considering that. Okay.
